### PR TITLE
[GPU] Fix get_default_params & choose_impl not to dependent on program_node 

### PR DESCRIPTION
**Pull request by yeonbok**
### Details:
 - This is baseline work for enabling dynamic shape for GPU plugin
 - For implementing dynamic shape, we need to be able to select layout and kernel impl& create primitive_impl independent on the common program_node from each stream 

### Tickets:
- 80676

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/include/kernel_selector_helper.h [hunk 1]
```diff
@@ -106,6 +108,54 @@ kernel_selector::weights_tensor convert_weights_tensor(const layout& l, bool is_
...
+
+    bool compensation_term = false;
+    layout compensation_layout = layout(data_types::f32, format::any, tensor());
+
+    kernel_impl_params(program& _prog, std::shared_ptr<const primitive> _desc, size_t _uid,
+                       const std::vector<layout>& _int_layouts, layout _out_layout,
+                       const std::vector<cldnn::fused_primitive_desc>& _fused_descs,
+                       const std::vector<activation_func>& _fused_act_funcs, const std::vector<activation_additional_params>& _act_params,
+                       layout _weights_layout = layout(data_types::f32, format::any, tensor()),
+                       bool   _bias_term = false,                    layout _bias_layout = layout(data_types::f32, format::any, tensor()),
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/include/kernel_selector_helper.h:141]: We have an `optional_data_type` wrapper for similar usage scenarios, so maybe it makes sense to introduce `optional_layout` which will combine presence flag with the data itself. 
Ideally, we need a template class similar to std::optional with some helpers on top of that. And if we transition from c++11 to c++17 at some point, we'll just replace custom implementation with the standard one.

File: src/plugins/intel_gpu/src/graph/include/program_node.h [hunk 2]
```diff
@@ -136,6 +125,17 @@ struct program_node {
     virtual primitive_type_id type() const { return desc->type; }
     virtual std::shared_ptr<kernel_selector::fuse_params> get_fuse_params() const { return nullptr; }
 
+    virtual std::shared_ptr<kernel_impl_params> get_kernel_impl_params() const {
+        return get_kernel_impl_params(get_input_layouts(), output_layout);
+    }
+
+    virtual std::shared_ptr<kernel_impl_params> get_kernel_impl_params(const std::vector<layout>& in_layouts,
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/include/program_node.h:132]: Do we need to return `shared_ptr` here? Can it be replaced with `unique_ptr`?

File: src/plugins/intel_gpu/src/graph/program_node.cpp [hunk 3]
```diff
@@ -897,12 +898,12 @@ void program_node::init_onednn_primitive_attributes() {
 
                 update_onednn_post_op_list(onednn_post_op_type::eltwise_act, empty_mem);
             }
-        } else if (node->is_type<eltwise>()) {
-            auto& e_node = node->as<eltwise>();
-            auto dep_idx = cldnn_post_ops[idx].dep_start_idx;
+        } else if (desc.is_type<eltwise>()) {
+            //auto& e_node = node->as<eltwise>();
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/program_node.cpp:902]: Please remove dead code from this file and some others.

File: src/plugins/intel_gpu/src/graph/include/program_node.h (hunk 2 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/include/program_node.h:132]: I thinks its possible. But will we get more benefit?

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/graph/include/program_node.h:132]: It's just a matter of clear design. If we don't really have shared ownership for these objects, then `unique_ptr` would be better.

File: src/plugins/intel_gpu/src/graph/include/kernel_selector_helper.h (hunk 1 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/include/kernel_selector_helper.h]: Good point, I applied it!

File: src/plugins/intel_gpu/src/graph/include/program_node.h (hunk 2 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/include/program_node.h]: Fixed.

File: src/plugins/intel_gpu/src/graph/program.cpp [hunk 4]
```diff
@@ -1039,10 +1039,12 @@ void program::fuse_nodes(program_node &fused_node,
                          program_node &peer_node,
                          std::map<primitive_id, std::vector<std::pair<primitive_id, size_t>>>* fusing_history) {
     auto peer_layout = peer_node.get_output_layout();
-    fused_primitive_desc local_desc;
-    local_desc.node = get_node_ptr(peer_node.id());
+    fused_primitive_desc local_desc(peer_node.get_primitive());
+    //local_desc.node = get_node_ptr(peer_node.id());
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/program.cpp:1043]: dead code

File: src/plugins/intel_gpu/src/graph/program_node.cpp [hunk 5]
```diff
@@ -870,14 +870,15 @@ void program_node::init_onednn_primitive_attributes() {
 
     int32_t num_sum_post_ops = 0;
     for (size_t idx = 0; idx < cldnn_post_ops.size(); idx++) {
-        auto node = cldnn_post_ops[idx].node;
+        //auto node = cldnn_post_ops[idx].node;
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/program_node.cpp:873]: dead code

File: src/plugins/intel_gpu/src/graph/include/kernel_selector_helper.h [hunk 6]
```diff
@@ -106,6 +108,48 @@ kernel_selector::weights_tensor convert_weights_tensor(const layout& l, bool is_
...
+    const program& prog;
+    std::shared_ptr<const primitive> desc;
+    size_t unique_id;
+    std::vector<layout> input_layouts;
+    layout output_layout;
+    std::vector<cldnn::fused_primitive_desc> fused_desc;
+    std::vector<activation_func> fused_act_funcs;
+    std::vector<activation_additional_params> activation_params;
+
+    //optional_layout weights_layout = layout(data_types::f32, format::any, tensor());
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/include/kernel_selector_helper.h:122]: dead code

File: src/plugins/intel_gpu/src/graph/impls/ocl/non_max_suppression.cpp [hunk 7]
```diff
@@ -101,21 +102,36 @@ struct non_max_suppression_impl : typed_primitive_impl_ocl<non_max_suppression>
...
+        auto get_second_output_node_idx = [&] {
+            size_t offset = 2;
+            offset += arg.has_num_select_per_class();
+            offset += arg.has_iou_threshold();
+            offset += arg.has_score_threshold();
+            offset += arg.has_soft_nms_sigma();
+            return offset;
+        };
+
+        auto get_third_output_node_idx = [&] {
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/impls/ocl/non_max_suppression.cpp:118]: These lambdas have pretty much the same code. Maybe single lambda with parameter would be better?
```cpp
auto get_output_node_idx = [&](size_t output_idx) {...}
```

**Review by vladimir-paramuzov — APPROVED** [review]
In general, LGTM

File: src/plugins/intel_gpu/src/graph/impls/ocl/non_max_suppression.cpp (hunk 7 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/impls/ocl/non_max_suppression.cpp]: Fixed, thanks!

File: src/plugins/intel_gpu/src/graph/include/kernel_selector_helper.h (hunk 6 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/include/kernel_selector_helper.h]: Fixed, thanks!

File: src/plugins/intel_gpu/src/graph/program_node.cpp (hunk 5 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/program_node.cpp]: Fixed!

**Pull request merged**